### PR TITLE
Add support for Nix

### PR DESCRIPTION
**Pull request by thblt**
This is a follow-up to #10 which adds support for running `build-step`s in a `nix-shell` environment.  It is made of two commits:

 - The first commit decouples the build-steps logic from `borg-build` and replace it with a `borg-run-build-steps-shell` function, which can be replaced by the user by setting the new variable `borg-run-build-steps-function`.  It also introduces an helper function `borg-build-step` which takes a string and an optional function (defaulting to `shell-command`) and executes the string through this function unless it looks like elisp (starts with a "("), in which case in `(eval)`uates it.  This is meant to keep the logic consistent between the default handler, provided, and user-provided handlers.

 - The second commit adds a `borg-run-build-steps-nix-shell` function, which wraps build-steps into a `nix-shell` configured through `submodules.DRONE.build-nix-shell-file` or `submodules.DRONE.build-nix-shell-packages`.  If none of these keys are set, it tries to automatically determine the best way to run the shell by checking if `shell.nix` or `default.nix` are present, in which case it passes no extra parameters, otherwise it just add an empty package list with `-p`.

I hope my Elisp isn't too bad, and reviewing it won't take you too much time.

Thanks

### CODE REVIEW

File: borg.el [hunk 1]
```diff
@@ -351,6 +351,31 @@ This function is to be used only with `--batch'."
...
+                             (or
+                              (file-exists-p (expand-file-name "default.nix" default-directory))
+                              (file-exists-p (expand-file-name "shell.nix" default-directory)))
+                           "-p")
+                         "")))
+    (when build
+      (message "Building %s from nix-shell %s" drone nix-shell-args)
+      (dolist (cmd build)
+        (borg-build-step cmd
+                         (lambda (cmd) (shell-command (concat "nix-shell --run " cmd " " nix-shell-args))))))))
```
Comment (tarsius) [inline borg.el]: Wouldn't we end up with something like `nix-shell -run make all -p file`? How would `nix-shell` know that `all` is an argument to `make` while `-p file` are arguments to itself? Maybe this should be `nix-shell -run 'make all' -p file`? Even then I would prefer `nix-shell -run -p file 'make all'` if that is allowed.

**Comment by tarsius** [comment]
I pushed two fixup commits to your branch.

**Comment by tarsius** [comment]
Wouldn't this achieve the same?

```diff
diff --git a/borg.el b/borg.el
index 80e50ec..b44ecb9 100644
--- a/borg.el
+++ b/borg.el
@@ -342,6 +342,8 @@ (defun borg-batch-rebuild-init ()
           (message "\n--- [%s] ---\n" f)
           (byte-recompile-file (expand-file-name f) t 0))))))
 
+(defvar borg-build-command nil "TODO")
+
 (defun borg-build (drone &optional activate)
   "Build the drone named DRONE.
 Interactively, or when optional ACTIVATE is non-nil,
@@ -349,13 +351,18 @@ (defun borg-build (drone &optional activate)
   (interactive (list (completing-read "Build drone: " (borg-drones) nil t)
                      t))
   (let ((default-directory (borg-worktree drone))
+        (build-cmd (and borg-build-command (funcall borg-build-command)))
+        ;; => e.g. "nix-shell -p file --run '%s'"
         (build (borg-get-all drone "build-step")))
     (if  build
         (dolist (cmd build)
           (message "  Running '%s'..." cmd)
-          (if (string-match-p "\\`(" cmd)
-              (eval (read cmd))
-            (shell-command cmd))
+          (cond ((string-match-p "\\`(" cmd)
+                 (eval (read cmd)))
+                (build-cmd
+                 (shell-command (format build-cmd cmd)))
+                (t
+                 (shell-command cmd)))
           (message "  Running '%s'...done" cmd))
       (let ((path (mapcar #'file-name-as-directory (borg-load-path drone))))
         (if noninteractive
```

File: borg.el (hunk 1 above)
Comment (thblt) [inline borg.el]: Indeed, my mistake, apologies. 

> Maybe this should be nix-shell -run 'make all' -p file? 

I think this is the only valid option.  That would give us something like:
``` elisp
(lambda (cmd)
  (shell-command (concat "nix-shell --run " (shell-quote-argument cmd) " " nix-shell-args))))))))
``` 

What do you think?

> Even then I would prefer nix-shell -run -p file 'make all' if that is allowed.

This can't be done.  The command must follow --run, and -p takes any number of arguments.

Comment (tarsius) [inline borg.el]: In summary, this works `nix-shell -run 'CMD' -p a b c`, right?

Comment (thblt) [inline borg.el]: Yes. -p is optional, if absent, the  path to a nix file describing the shell environment can be passed as the last parameter. If neither -p nor a path are provided, it will look for a shell.nix or a default.nix file in the current directory.

**Comment by thblt** [comment]
Yes, this should work, but we need to pass the drone name to `borg-build-command` and I'm not sure about the use of `(format)`: in the nix-shell case, how is it possible to wrap the cmd in `(shell-quote-argument)`?

Your approach is obviously cleaner than mine,but I think it would be better if `borg-build-function` was a function taking a drone name and returning another function taking a string and returning it correctly wrapped. This would also allow to simplify borg-build a bit by giving it a simple default value:

    (defvar borg-build-command (lambda (x) (lambda (y) y)) "TODO")

So there's no more need to test if it's non-nil.

Sorry about formatting, I'm on mobile and won't be back near a computer until Saturday.

**Comment by thblt** [comment]
Working on this right now.  I'm replacing `(if build ...` by `(when build ...`, I hope it's not a mistake.  As I understand it, the fact that a package requires special build steps doesn't mean these steps will make the "standard" Borg build (autoloads, byte compile, makeinfo) unnecessary.  Please correct me if I'm wrong.

**Comment by thblt** [comment]
This commit implements your suggestion.  It adds a:

``` elisp
(defvar borg-build-command
  (lambda (_) (lambda (y) y))
  ...)
```

and provides an implementation of a nix-shell wrapper in ```borg-nix-shell-build-command```.  It also now defaults to passing a `--pure` flag to nix-shell, which completely isolates the shell from the environment, and provides a global option to disable it.

**Comment by tarsius** [comment]
> As I understand it, the fact that a package requires special build steps doesn't mean these steps will make the "standard" Borg build (autoloads, byte compile, makeinfo) unnecessary. Please correct me if I'm wrong.

For some packages using the upstream build tool and instructions makes all of Borg's build steps redundant, while for other packages it only makes some of them redundant. If the upstream way of building the package makes some of Borg's own build steps redundant, then we want to skip them. Because of that a package that does set `build-step` to build as intended by upstream may have to specify additional `build-step`s that call only those of the Borg's default build-steps that are still required.

For example `org` takes care of everything so there is just one `build-step`:

```
[submodule "org"]
	path = lib/org
	url = git://orgmode.org/org-mode.git
	build-step = make
	load-path = lisp
	load-path = contrib/lisp
	info-path = doc
```

It would be a waste to call `borg-update-autoloads`, `borg-byte-compile`, and `borg-makeinfo` for this package.

But `mu4e` does not generate the autloads file, so `borg-update-autoloads` has to be called.

```
[submodule "mu4e"]
	path = lib/mu4e
	url = git@github.com:djcb/mu.git
	build-step = test -e ./configure || autoreconf -i
	build-step = ./configure
	# When first installing mu and after major changes you have to
	# build it using "make".  If you prefer you can do that every
	# time mu/mu4e has new commits , but that is much slower, and
	# usually not necessary.
	# build-step = make > /dev/null
	build-step = make -C mu4e > /dev/null
	build-step = "(borg-update-autoloads \"mu4e\")"
	load-path = mu4e
```

**Comment by thblt** [comment]
Thanks again for the detailed clarification. I reverted the original behavior in the last commit.

**Comment by tarsius** [comment]
> This commit implements your suggestion.

Not really. I consider the change I posted above to be the complete implementation (except for documentation and maybe renaming the variable). The idea is that you implement your `nix` function on top of that. I have now pushed this to the `build-command` branch.

Your implementation is much more complex and much more code and I fail to see what benefit it has over mine.

Also I don't think that `borg-nix-shell-build-command` should be added to `borg.el`, instead it should be implemented in a separate package, likely named `borg-nix`. It would be nice if that also defined `borg-guix-shell-build-command`.

**Comment by tarsius** [comment]
Ah, I didn't see you made significant changes before I posted the above comment. Will take another look.

**Comment by tarsius** [comment]
So now your version basically only differs with regards to the function returning function thing. That's an interesting idea, I'll consider adding it. But for your use-case that isn't really necessary is it.

I don't think that putting your function in a separate package would cause bootstrapping issues, but if it does, then that would be a good argument for adding to `borg.el` itself.

**Comment by thblt** [comment]
> So now your version basically only differs with regards to the function returning function thing. (...) But for your use-case that isn't really necessary, is it?

Not for Nix itself, but I'm not sure all possible build wrappers would take a shell-quoted string.  For example, the [Guix documentation](https://www.gnu.org/software/guix/manual/html_node/Invoking-guix-environment.html) *seems* to indicate that `nix-shell --run "make all"` would translate to `guix environment -- make all` --- unquoted.   I haven't tested, but if this is actually the case, a function is needed (or at least a parameter to determine if the build-step should be quoted)

What I really can't find a use case for is the fact that the function returned by `(funcall borg-build-command drone)` can return nil, in which case nothing is done with the result and the function is assumed to have been run for effects only.  I assume it can be a nice thing to have to handle the weirder cases.

Another small difference is that in my version, `borg-build-command`  takes an explicit `(drone)` argument, but this was me not understanding Lisp name scoping.  

> I don't think that putting your function in a separate package would cause bootstrapping issues

That should be fine as long as Borg uses the default build mechanism for itself, as it currently does.

**Comment by tarsius** [comment]
I have merged an alternative implementation, see https://github.com/emacscollective/borg/issues/10#issuecomment-323821348.

**Comment by thblt** [comment]
Thank you!
